src/query/query_hash_scan.c

```diff
@@ -0,0 +1,441 @@
+      /* make a copy */
+      new_key = qdata_alloc_hscan_key (thread_p, key->val_count, false);
+    }
+
+  if (new_key)
+    {
+      /* copy values */
+      new_key->val_count = key->val_count;
+      for (i = 0; i < key->val_count; i++)
+	{
+	  vtype1 = REGU_VARIABLE_GET_TYPE (&probe_regu_list->value);
+	  vtype2 = DB_VALUE_DOMAIN_TYPE (key->values[i]);
+
+	  if (vtype1 != vtype2)
+	    {
+	      new_key->values[i] = pr_make_value ();
+	      if (new_key->values[i] == NULL)
+		{
+		  new_key->free_values = true;
+		  qdata_free_hscan_key (thread_p, new_key, i);
+		  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (DB_VALUE *));
+		  return NULL;
+		}
+
+	      status = tp_value_coerce (key->values[i], new_key->values[i], probe_regu_list->value.domain);
+	      if (status != DOMAIN_COMPATIBLE)
+		{
+		  new_key->free_values = true;
+		  qdata_free_hscan_key (thread_p, new_key, ++i);
+		  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TP_CANT_COERCE, 2, pr_type_name (vtype2),
+			  pr_type_name (vtype1));
+		  return NULL;
+		}
+	    }
+	  else
+	    {
+	      new_key->values[i] = pr_copy_value (key->values[i]);
+	    }
+	  probe_regu_list = probe_regu_list->next;
+	}
+
+      new_key->free_values = true;
+    }
+
+  return new_key;
+}
+
+/*
+ * qdata_alloc_agg_hkey () - allocate new hash aggregate key
+ *   returns: pointer to new structure or NULL on error
+ *   thread_p(in): thread
+ */
+HASH_SCAN_VALUE *
+qdata_alloc_hscan_value (cubthread::entry * thread_p, QFILE_TUPLE tpl)
+{
+  HASH_SCAN_VALUE *value;
+  int tuple_size = QFILE_GET_TUPLE_LENGTH (tpl);
+
+  /* alloc structure */
+  value = (HASH_SCAN_VALUE *) db_private_alloc (thread_p, sizeof (HASH_SCAN_VALUE));
+  if (value == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (HASH_SCAN_VALUE));
+      return NULL;
+    }
+
+  value->tuple = (QFILE_TUPLE) db_private_alloc (thread_p, tuple_size);
+  if (value->tuple == NULL)
+    {
+      qdata_free_hscan_value (thread_p, value);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, tuple_size);
+      return NULL;
+    }
+  /* save tuple */
+  if (!safe_memcpy (value->tuple, tpl, tuple_size))
+    {
+      qdata_free_hscan_value (thread_p, value);
+      return NULL;
+    }
+  return value;
+}
+
+static bool
+safe_memcpy (void *data, void *source, int size)
+{
+  if (size < 0)
+    {
+      return false;
+    }
+  memcpy (data, source, (size_t) size);
+  return true;
+}
+
+/*
+ * qdata_free_hscan_value () - free hash value
+ *   thread_p(in): thread
+ *   key(in): hash value
+ */
+void
+qdata_free_hscan_value (cubthread::entry * thread_p, HASH_SCAN_VALUE * value)
+{
+  if (value == NULL)
+    {
+      return;
+    }
+
+  /* free values */
+  if (value->tuple != NULL)
+    {
+      db_private_free (thread_p, value->tuple);
+      value->tuple = NULL;
+    }
+  /* free structure */
+  db_private_free (thread_p, value);
+  value = NULL;
```

Comment:
```suggestion
  db_private_free_and_init (thread_p, value->tuple);
  
  /* free structure */
  db_private_free_and_init (thread_p, value);
```
it is just a minor suggestion.

Reply:
I fixed it